### ISSUE TITLE
Xml conformance: CR and CRLF should be normalized to LF

### ISSUE THREAD

**Issue by corliss** [AngleSharp/AngleSharp#149]
I'm trying to use `XmlTokenizer` as a replacement for `XmlReader`. Apart from the separately discussed performance issue, here is one possible conformance issue.

According to the spec (http://www.w3.org/TR/2008/REC-xml-20081126/#sec-line-ends): "To simplify the tasks of applications, the XML processor must behave as if it normalized all line breaks in external parsed entities (including the document entity) on input, before parsing, by translating both the two-character sequence #xD #xA and any #xD that is not followed by #xA to a single #xA character."

`XmlReader` does follow this spec, but `XmlTokenizer` appears to not follow it.

Do you agree that this is an issue? It's the only thing blocking me from moving to AngleSharp Xml at present.

**Comment by FlorianRappl** [comment]
Hm, really? How do you come to that conclusion? All CRLF are normalized to LF in the `BaseTokenizer`, which is used by all tokenizers.

(Also single CR are converted to LF)

**Comment by corliss** [comment]
I have some unit tests that check for this, that are failing with `XmlTokenizer`. I will check and let you know - maybe it's a problem in the test code.

**Comment by corliss** [comment]
OK it looks like a slightly different issue, but it might be in AngleSharp.
Given this input: `<pre>\n \t\r</pre>`
The tokenizer emits:

```
StartTag: pre
Character: \n
Character: [space]
Character: \t
EndTag
```

In other words, the last `\r` is eaten.

I'll still poke around to see if this is in my code, but at this point I do suspect it might be in AngleSharp. Does this ring any bells?

**Comment by corliss** [comment]
In BaseTokenizer.cs:

``` C#
        void AdvanceUnsafe()
        {
            if (_current.IsLineBreak())
            {
                _columns.Push(_column);
                _column = 0;
                _row++;
            }

            _column++;
            _current = _source.ReadCharacter();

            while (_current == Symbols.CarriageReturn)
            {
                _current = _source.ReadCharacter();
            }
        }

```

All carriage returns are being ignored, whereas only a carriage return that is followed by a newline should be ignored.

Do you agree?

**Comment by corliss** [comment]
Please do let me know what you think. If it is a bug in AngleSharp perhaps I can use a temporary workaround (or even better, a fix - if it is easy to implement.)

That way I will be unblocked to switch over to using AngleSharp's XML parser.

**Comment by FlorianRappl** [comment]
Your code is not up to date. Please update it! The snippet you show in the `BaseTokenizer` has been changed with v0.9.1.
(has been discussed in #142)

**Comment by corliss** [comment]
I see. Unfortunately I am kind of unable to update because of #145. My code still targets the earlier `ToCss()` methods. Any chance of #145 being fixed in the near future? (At this point it looks like a fairly small set of changes.)

**Comment by corliss** [comment]
I was able to update to the latest AngleSharp and indeed this issue is fixed (but there's another issue I've opened separately.) Thanks!
